aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,354 @@
+        let buckets = table_with_length::new();
+        table_with_length::add(&mut buckets, 0, vector::empty());
+        let table = SmartTable {
+            buckets,
+            num_buckets: 1,
+            level: 0,
+            size: 0,
+            // The default split load threshold is 75%.
+            split_load_threshold: if (split_load_threshold == 0) {75} else {split_load_threshold},
+            target_bucket_size,
+        };
+        // The default number of initial buckets is 2.
+        if (num_initial_buckets == 0) {
+            num_initial_buckets = 2;
+        };
+        while (num_initial_buckets > 1) {
+            num_initial_buckets = num_initial_buckets - 1;
+            split_one_bucket(&mut table);
+        };
+        table
+    }
+
+    /// Destroy empty table.
+    /// Aborts if it's not empty.
+    public fun destroy_empty<K, V>(table: SmartTable<K, V>) {
+        assert!(table.size == 0, error::invalid_argument(ENOT_EMPTY));
+        let i = 0;
+        while (i < table.num_buckets) {
+            vector::destroy_empty(table_with_length::remove(&mut table.buckets, i));
+            i = i + 1;
+        };
+        let SmartTable {buckets, num_buckets: _, level: _, size: _, split_load_threshold:_, target_bucket_size: _} = table;
+        table_with_length::destroy_empty(buckets);
+    }
+
+    /// Destroy a table completely when V is dropable.
+    public fun destroy<K: drop, V: drop>(table: SmartTable<K, V>) {
+        let i = 0;
+        while (i < table.num_buckets) {
+            let j = 0;
+            let bucket = table_with_length::remove(&mut table.buckets, i);
+            let bucket_length = vector::length(&bucket);
+            while (j < bucket_length) {
+                vector::pop_back(&mut bucket);
+                j = j + 1;
+            };
+            vector::destroy_empty(bucket);
+            i = i + 1;
+        };
+        let SmartTable {buckets, num_buckets: _, level: _, size: _, split_load_threshold:_, target_bucket_size: _} = table;
+        table_with_length::destroy_empty(buckets);
+    }
+
+    /// Add (key, value) pair in the hash map, it may grow one bucket if current load factor exceeds the threshold.
+    /// Note it may not split the actual overflowed bucket. Instead, it was determined by `num_buckets` and `level`.
+    /// For standard linear hash algorithm, it is stored as a variable but `num_buckets` here could be leveraged.
+    /// Abort if `key` already exists.
+    /// Note: This method may occasionally cost much more gas when triggering bucket split.
+    public fun add<K, V>(table: &mut SmartTable<K, V>, key: K, value: V) {
+        let hash = sip_hash_from_value(&key);
+        let index = bucket_index(table.level, table.num_buckets, hash);
+        let bucket = table_with_length::borrow_mut(&mut table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
```

Comment:
You can use vector::for_each_ref here. Can you check all while loops and see if you can replace with inline functions?